lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -124,11 +120,30 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
     -- then you can skip requesting the headers and go straight to requesting
     -- the block itself.
     blocks <- if singleBlockHeader
-              then requestBlocks (NewestFirst (pure tipHeader))
-              else requestHeaders >>= requestBlocks
+              then requestBlocks (OldestFirst (one tipHeader))
+              else requestAndClassifyHeaders >>= requestBlocks
     pure (OldestFirst (reverse (toList blocks)))
   where
 
+    requestAndClassifyHeaders :: d (OldestFirst NE BlockHeader)
+    requestAndClassifyHeaders = do
+        headers <- toOldestFirst <$> requestHeaders
+        getLcaMainChain logic headers >>= \case
+            Nothing -> throwM $ DialogUnexpected $ "Got headers, but couldn't compute " <>
+                                                   "LCA to ask for blocks"
+            Just (lca :: HeaderHash) -> do
+                -- Headers list is (oldest to newest)
+                -- [n1,n2,...nj,lca,nj+2,...nk] we drop [n1..lca] and
+                -- return [nj+2..nk], as we already have lca in our
+                -- local db. Usually this function does 1 iterations
+                -- as it's a common case that [n1..nj] is absent and
+                -- lca is the oldest header.
+                let dropUntilLca = NE.dropWhile (\h -> h ^. prevBlockL /= lca)
+                case nonEmpty (dropUntilLca $ getOldestFirst headers) of
+                    Nothing -> throwM $ DialogUnexpected $
+                                   "All headers are older than LCA, nothing to query"
```

Comment:
I'm not demanding this change here but I just want to ask, do we agree that this would be better as an explicit `Left (DialogUnexpected ...)`? So that the type of `getBlocks` would end in `d (Either <error type> OldestFirst [] Block)`?

Reply:
This should be a completely different story, I think. Currently, `retrievalWorker` expects an exception as a sign of "something went wrong", so we eventually throw even if we get `Left` somewhere. So at this point, no. But I'd took a look at exception handling in retrieval worker. 

The suggestion itself is something to think about. At a logic layer, I don't really care whether it was a "normal" exception like "somebody sent us trash" or "exceptional" like "my computer is on fire". But maybe I'm missing something.